### DISCUSSION TITLE
Misconception about async atoms, getDefaultStore and unwrap..

### DISCUSSION THREAD

**Discussion by btmnk** [pmndrs/jotai#2249] [Q&A] [answered]
In my project I use an async atom to load my app config from a config.json. If I understood everything correctly React suspends the app until the atom is resolved. This way my app will only render as soon as the config is available.
So far so good.

Now I do have a use case where I want to use that config synchronously without a hook (because it would be way more overhead to pass down a callback from a hook).

My understanding was that once the async atom resolved at least once the data is already written into the store. Meaning as soon as React lifts the suspense the config should be available globally in the default store? Right?

The problem I'm facing is, that whenever I use `getDefaultStore().get(unwrap(appConfigAtom))` to get that config, it's always undefined when called immediately with a render. When using it on a button click or with a delay it's fine.

I also checked by logging the timestamps when unwrapping and when awaiting and indeed it seems the Promise actually takes a few ms before resolving (even though the config.json has already been loaded minutes ago and other components already used the config correctly with the useAtom hook).

Am I misunderstanding how things work internally with async atoms?

//EDIT:
https://codesandbox.io/s/jotai-unwrap-async-fq8hm9?file=/src/AppConfig.ts
Code sandbox showing my use case

**Comment by dai-shi** [comment]
unwrapped atom returns `undefined` in the loading state.

> that once the async atom resolved at least once the data is already written into the store.

So, that's the misconception.  Jotai doesn't hold previous values.
You can use unwrap's second argument to customize the behavior. This is a workaround because using a previous value means the atom model isn't 100% pure functions.

```js
unwrap(appConfigAtom, (prev) => prev)
```

**Reply by btmnk** [reply]
But when using the synchronous call it doesn't seem to reexecute the promise inside the atom. So I assumed it isn't in the loading state anymore (since the atom already resolved otherwise the page wouldn't even show because of the Suspense). 

> So, that's the misconception. Jotai doesn't hold previous values.

But it wouldn't be the previous value, it would be the current one.. no? I only need the one value that resolved when the atom was called the first time with a hook while the app was suspended.

Using `unwrap(appConfigAtom, (prev) => prev)` also just always returns undefined for me.

I thought regardless when and how I try to access an atom it will always return the same reference (in this case to the Promise) and once that is resolved it is resolved for every access. With unwrap I just tell the API that I know it will be resolved when this gets called (since the app couldn't be visible unless it already resolved) and I get the resolved value from the Promise (reference to the very same Promise used in the hooks). Is this wrong? Are hooks and `store.get()` calls using different "versions" of the referenced Promise?

**Reply by dai-shi** [reply]
> But when using the synchronous call it doesn't seem to reexecute the promise inside the atom.

Hmm, it sounds like I misunderstood something. Sorry about that. Would you be able to create a very small reproduction with https://csb.jotai.org ?

**Reply by btmnk** [reply]
https://codesandbox.io/s/jotai-unwrap-async-fq8hm9?file=/src/AppConfig.ts

**Reply by btmnk** [reply]
I'm also wondering right now if I might have misunderstood how Suspense works.
It could simply be that in the first render everything gets executed once before React notices that useAppConfig needs to Suspend the App. So the call with getAppConfig was already executed parallel to the hook.

Once the Suspense is lifted I think React does not remount everything so it keeps the previously rendered state...

**Reply by dai-shi** [reply, marked as answer]
I see. What you observe is because when you read unwrapped atom, it's not initialized because it's the first use, so it returns `undefined`. It will be updated soon, but you need to subscribe to the atom to get notified. So, the correct conception is atom is lazy evaluated for each.

**Reply by btmnk** [reply]
I think I understand now, thank you for your time and the explanations! Accessing the atom synchronously on the first render would have been nice but I solved it for now by using `useAtom` and passing the value down to the functions that need it.

**Reply by dai-shi** [reply]
This might be interesting to you: https://twitter.com/dai_shi/status/1722543472874594334
